extmod/modbluetooth.c

```diff
@@ -0,0 +1,1002 @@
+    mp_int_t conn_handle = mp_obj_get_int(conn_handle_in);
+    return bluetooth_handle_errno(mp_bluetooth_gattc_discover_primary_services(conn_handle));
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_2(bluetooth_ble_gattc_discover_services_obj, bluetooth_ble_gattc_discover_services);
+
+STATIC mp_obj_t bluetooth_ble_gattc_discover_characteristics(size_t n_args, const mp_obj_t *args) {
+    mp_int_t conn_handle = mp_obj_get_int(args[1]);
+    mp_int_t start_handle = mp_obj_get_int(args[2]);
+    mp_int_t end_handle = mp_obj_get_int(args[3]);
+    return bluetooth_handle_errno(mp_bluetooth_gattc_discover_characteristics(conn_handle, start_handle, end_handle));
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(bluetooth_ble_gattc_discover_characteristics_obj, 4, 4, bluetooth_ble_gattc_discover_characteristics);
+
+STATIC mp_obj_t bluetooth_ble_gattc_discover_descriptors(size_t n_args, const mp_obj_t *args) {
+    mp_int_t conn_handle = mp_obj_get_int(args[1]);
+    mp_int_t start_handle = mp_obj_get_int(args[2]);
+    mp_int_t end_handle = mp_obj_get_int(args[3]);
+    return bluetooth_handle_errno(mp_bluetooth_gattc_discover_descriptors(conn_handle, start_handle, end_handle));
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(bluetooth_ble_gattc_discover_descriptors_obj, 4, 4, bluetooth_ble_gattc_discover_descriptors);
+
+STATIC mp_obj_t bluetooth_ble_gattc_read(mp_obj_t self_in, mp_obj_t conn_handle_in, mp_obj_t value_handle_in) {
+    mp_int_t conn_handle = mp_obj_get_int(conn_handle_in);
+    mp_int_t value_handle = mp_obj_get_int(value_handle_in);
+    return bluetooth_handle_errno(mp_bluetooth_gattc_read(conn_handle, value_handle));
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_3(bluetooth_ble_gattc_read_obj, bluetooth_ble_gattc_read);
+
+STATIC mp_obj_t bluetooth_ble_gattc_write(size_t n_args, const mp_obj_t *args) {
+    mp_int_t conn_handle = mp_obj_get_int(args[1]);
+    mp_int_t value_handle = mp_obj_get_int(args[2]);
+    mp_obj_t data = args[3];
+    mp_buffer_info_t bufinfo = {0};
+    mp_get_buffer_raise(data, &bufinfo, MP_BUFFER_READ);
+    size_t len = bufinfo.len;
+    return bluetooth_handle_errno(mp_bluetooth_gattc_write(conn_handle, value_handle, bufinfo.buf, &len));
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(bluetooth_ble_gattc_write_obj, 4, 4, bluetooth_ble_gattc_write);
+
+#endif // MICROPY_PY_BLUETOOTH_ENABLE_CENTRAL_MODE
+
+// ----------------------------------------------------------------------------
+// Bluetooth object: Definition
+// ----------------------------------------------------------------------------
+
+STATIC const mp_rom_map_elem_t bluetooth_ble_locals_dict_table[] = {
+    // General
+    { MP_ROM_QSTR(MP_QSTR_active), MP_ROM_PTR(&bluetooth_ble_active_obj) },
+    { MP_ROM_QSTR(MP_QSTR_config), MP_ROM_PTR(&bluetooth_ble_config_obj) },
+    { MP_ROM_QSTR(MP_QSTR_irq), MP_ROM_PTR(&bluetooth_ble_irq_obj) },
+    // GAP
+    // TODO: Potentially rename these to gap_*.
```

Comment:
I would prefer `gap_connect` etc, just to be consistent with the `gatts_` / `gattc_` of the other methods. What do you think?

Reply:
gattc/gatts are necessary because of the C vs S; gap is not needed from this point of view, the names like "connect" are unique.  But, considering the low-level, BLE-spec-following nature of the API, it does make sense to use `gap_connect`, `gap_advertise` etc.  It's just a lot of characters in all the corresponding qstrs...

Reply:
I've switched it to `gap_`. 16 extra bytes of qstrs. But... 400 bytes saved by collapsing all the connect/disconnect event paths down to one, now that disconnect also includes the address.